### PR TITLE
Feature/parallelization

### PR DESCRIPTION
**Pull request by Jibbarth**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Fixed tickets | #390 

This PR add parallelization feature. 

No external dependencies added, I create a new hidden command 'internal:processors' that launch code sniffer and CS fixer on a list of files.  

The runner create as much as it can be new processes for this new command by retrieving number of CPU core, and launch them in parallel.

The new command store all details and fixed issues in cache, then the runner load issues from cache.

Please note that is do only for CodeSniffer and CSFixer processors. 

I cannot add Insights instances in cache because there is a problem to cache SplFileInfo (not serializable), so each process need to reinstance them. 

Let's talk about performance now. 

* Performance on Master without Cache : https://blackfire.io/profiles/c161aa31-3081-449c-b359-e481eb2be04c/graph => ~ 8s
* Performance on Master with Cache : https://blackfire.io/profiles/793dd7c1-0fcb-44aa-a564-731f02818f3e/graph => ~ 1.24s
* Performance on this branch without cache : https://blackfire.io/profiles/60f40fe3-c798-404e-a1e3-35d0f8cb34e3/graph => ~ 2,12s
* Performance on this branch with cache : https://blackfire.io/profiles/bd5f7ac3-060b-4671-af73-41607ff7cc2f/graph => ~ 1.56s

Gain without cache : https://blackfire.io/profiles/compare/21eebd42-e83c-4bb4-99ad-8e717f42e6e9/graph 
=> 
image

When cache is already set, I loose slight performance: https://blackfire.io/profiles/compare/6dd977d1-c140-4016-a536-f03274744eca/graph 

Maybe should we didn't launch subprocess in this case.
**Edit** : We bypass now launching subprocess for files that are already in cache :+1: 

--- 

TODO : 
- [x] Fix coding standards (phpstan etc.)
- [x] Add more tests ? 
- [x] Document the new `phpinsights.threads` option (in stubs)
- [x] Test on laravel application, launched with `artisan insights`
- [x] Test on Windows (if someone can do the test for me ? 🤗)
- [ ] ...

### CODE REVIEW

**Comment by nunomaduro** [comment]
@50bhan Wanna review this one?

**Comment by 50bhan** [comment]
@nunomaduro Sure, I'll check it out tomorrow.

**Comment by 50bhan** [comment]
Perfect work @Jibbarth :+1: Really liked the approach and implementation in clean as always. I have some comments though.

File: src/Domain/Runner.php [hunk 1]
```diff
@@ -71,111 +71,131 @@ public function addInsights(array $insights): void
...
+        $progressBar = $this->createProgressBar($totalFiles + \count($this->globalInsights));
         $progressBar->setMessage('');
         $progressBar->start();
 
-        /** @var GlobalInsight $insight */
-        foreach ($this->globalInsights as $insight) {
-            if ($this->output->isVerbose()) {
-                $progressBar->setMessage(' - Global: ' . $insight->getTitle());
-                $progressBar->display();
+        $this->cache->set('current_configuration', $this->configuration);
```
Comment (50bhan) [inline src/Domain/Runner.php:104]: Do we need to do this after we done it once already (line 89)?

**Comment by 50bhan** [comment]
Regarding performance, currently we are looking in cache for files at `InternalProcessorCommand` class. So if we move this check to `Runner` and before creation of processes, we don't need to create even useless processes. The code can be something like below image:

Screenshot from 2020-06-22 15-24-47

I did a test and you can check the result: https://blackfire.io/profiles/8dcc20f6-2435-4161-896e-92513fa2ece6/graph .

**Comment by Jibbarth** [comment]
@50bhan I had in mind to check if first file and last file had cache, but your idea is more clever :+1:

**Comment by Jibbarth** [comment]
Anyone who use Windows can help me to test this PR ?

It's pretty simple : 

### On an other project

Add this in `composer.json`

```bash
composer config repositories.phpinsights vcs https://github.com/jibbarth/phpinsights
```

Then require phpinsights on correct branch : 

```bash
composer require nunomaduro/phpinsights:dev-feature/parallelization
```

### On phpinsights repo 

```bash
git remote add jibbarth git@github.com:Jibbarth/phpinsights.git
git fetch jibbarth
git checkout feature/parallelization
```

**Comment by ekvedaras** [comment]
I can test tomorrow on a bigger project. Will reply with the results

**Comment by ekvedaras** [comment]
## Stable version

Finished finished in about 2 minutes

stable-version

## I was unable to run parallelized version though.

**In Process.php line 338:**

  [Symfony\Component\Process\Exception\RuntimeException]
  **Unable to launch a new process.**


**Exception trace:**
  ```
at C:\Users\ekvedaras\AppData\Roaming\Composer\vendor\symfony\process\Process.php:338
 Symfony\Component\Process\Process->start() at C:\Users\ekvedaras\AppData\Roaming\Composer\vendor\symfony\process\Process.php:232
 Symfony\Component\Process\Process->run() at C:\Users\ekvedaras\AppData\Roaming\Composer\vendor\nunomaduro\phpinsights\src\Domain\Insights\SyntaxCheck.php:52
 NunoMaduro\PhpInsights\Domain\Insights\SyntaxCheck->process() at C:\Users\ekvedaras\AppData\Roaming\Composer\vendor\nunomaduro\phpinsights\src\Domain\Runner.php:161
 NunoMaduro\PhpInsights\Domain\Runner->run() at C:\Users\ekvedaras\AppData\Roaming\Composer\vendor\nunomaduro\phpinsights\src\Domain\Insights\InsightFactory.php:98
 NunoMaduro\PhpInsights\Domain\Insights\InsightFactory->runInsightCollector() at C:\Users\ekvedaras\AppData\Roaming\Composer\vendor\nunomaduro\phpinsights\src\Domain\Insights\InsightFactory.php:69
 NunoMaduro\PhpInsights\Domain\Insights\InsightFactory->makeFrom() at C:\Users\ekvedaras\AppData\Roaming\Composer\vendor\nunomaduro\phpinsights\src\Domain\Insights\InsightCollectionFactory.php:67
 NunoMaduro\PhpInsights\Domain\Insights\InsightCollectionFactory::NunoMaduro\PhpInsights\Domain\Insights\{closure}() at n/a:n/a
 array_map() at C:\Users\ekvedaras\AppData\Roaming\Composer\vendor\nunomaduro\phpinsights\src\Domain\Insights\InsightCollectionFactory.php:68
 NunoMaduro\PhpInsights\Domain\Insights\InsightCollectionFactory->get() at C:\Users\ekvedaras\AppData\Roaming\Composer\vendor\nunomaduro\phpinsights\src\Application\Console\Analyser.php:36
 NunoMaduro\PhpInsights\Application\Console\Analyser->analyse() at C:\Users\ekvedaras\AppData\Roaming\Composer\vendor\nunomaduro\phpinsights\src\Application\Console\Commands\AnalyseCommand.php:65
 NunoMaduro\PhpInsights\Application\Console\Commands\AnalyseCommand->__invoke() at n/a:n/a
 call_user_func() at C:\Users\ekvedaras\AppData\Roaming\Composer\vendor\nunomaduro\phpinsights\src\Application\Console\Commands\InvokableCommand.php:34
 NunoMaduro\PhpInsights\Application\Console\Commands\InvokableCommand->execute() at C:\Users\ekvedaras\AppData\Roaming\Composer\vendor\symfony\console\Command\Command.php:255
 Symfony\Component\Console\Command\Command->run() at C:\Users\ekvedaras\AppData\Roaming\Composer\vendor\symfony\console\Application.php:912
 Symfony\Component\Console\Application->doRunCommand() at C:\Users\ekvedaras\AppData\Roaming\Composer\vendor\symfony\console\Application.php:264
 Symfony\Component\Console\Application->doRun() at C:\Users\ekvedaras\AppData\Roaming\Composer\vendor\symfony\console\Application.php:140
 Symfony\Component\Console\Application->run() at C:\Users\ekvedaras\AppData\Roaming\Composer\vendor\nunomaduro\phpinsights\bin\phpinsights:35
 {closure}() at C:\Users\ekvedaras\AppData\Roaming\Composer\vendor\nunomaduro\phpinsights\bin\phpinsights:36
```

**Comment by ekvedaras** [comment]
I guess that project is just too big 😀 Tried to dump the syntax check command and run it manually.

```
Program 'php.exe' failed to run: The filename or extension is too longAt line:1 char:1
+ C:\php\php.exe "C:\Users\ekvedaras\AppData\Roaming\Composer\vendor\nu ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
At line:1 char:1
+ C:\php\php.exe "C:\Users\ekvedaras\AppData\Roaming\Composer\vendor\nu ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ResourceUnavailable: (:) [], ApplicationFailedException
    + FullyQualifiedErrorId : NativeCommandFailed
```

**Comment by ekvedaras** [comment]
Might be the culprit https://stackoverflow.com/a/29359348/3805289 . There is a limit on command length. Even when I try to paste the command to windows terminal it already warns about the command being longer then 5KiB.

**Comment by Jibbarth** [comment]
@ekvedaras Thanks you so much for testing this ! 

If I understand, the bug came from the SyntaxCheck insight which was added in #373 ? Do you reproduce it on `dev-master` ? 

Can you exclude the insight `NunoMaduro\PhpInsights\Domain\Insights\SyntaxCheck` in your `phpinsights.php` config file and recheck ? 

You are probably right about command lenght, that's something we should care about :confused:

**Comment by ekvedaras** [comment]
I tested again on a smaller project.

Stable version:
stable-version-mid-project

Parallel version:
parallel-version-mid-project

Not a supper big difference. The cached version ran instantly of course :)) My laptop was not completely idle, but I tried to close most of the apps.
Btw, not sure if relevant. But I saw it fetches the number of cores as 4. Would it make sense to fetch threads instead which is 8 ? Although it seemed like all threads were receiving load, so maybe that works.

**Comment by Jibbarth** [comment]
@ekvedaras You are right, we used `wmic cpu get NumberOfCores` to retrieve number of threads available, but I found `wmic cpu get NumberOfLogicalProcessors` that return the correct result (same results as on my Linux)

I also reduce size of args by passing array of files to analyze by thread in the cache, so we should no longer had problem about command size.

**Comment by ekvedaras** [comment]
Tried to run it again. Seems a lot faster actually. But I'm hitting this issue now. Event with `--disable-security-check`
image

**Comment by ekvedaras** [comment]
**CPU**

	Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz

	Base speed:	2,80 GHz
	Sockets:	1
	Cores:	4
	Logical processors:	8



### v1.14.0

v1-14

Time: **1m 48s**

### Parallel

parallel-new-version

Time: **28s**

That's more like it 👍

**Review by austinheap — APPROVED** [review]

**Comment by Jibbarth** [comment]
@ekvedaras Thanks you so much for helping me to test this feature :pray: 

@50bhan @olivernybroe I think this PR is ready for review :innocent:

**Review by olivernybroe — APPROVED** [review]
Amazing work @Jibbarth!

Just tested it out locally and it worked perfectly fine on my mac.

I'll say this is ready now 🥇 

And thanks to @ekvedaras for helping out on this 👍

**Comment by 50bhan** [comment]
Great job @Jibbarth :+1: I think everything looks great and ready to merge!

**Comment by olivernybroe** [comment]
Alright.

@Jibbarth You got green light to merge it in 👍

**Comment by Jibbarth** [comment]
Thank you all :raised_hands:

**Pull request merged**